### ISSUE TITLE
Order of `id` column in `select` changes loaded record ID

### ISSUE THREAD

**Issue by conzett** [rails/rails#48944]
### Steps to reproduce

```ruby
require "bundler/inline"                                                         
                                                                                 
gemfile(true) do                                                                 
  source "https://rubygems.org"                                                  
                                                                                 
  git_source(:github) { |repo| "https://github.com/#{repo}.git" }                
                                                                                 
  # Activate the gem you are reporting the issue against.                        
  gem "activerecord", "~> 7.0.0"                                                 
  gem "sqlite3"                                                                  
end                                                                              
                                                                                 
require "active_record"                                                          
require "minitest/autorun"                                                       
require "logger"                                                                 
                                                                                 
# This connection will do for database-independent bug reports.                  
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)                                   
                                                                                 
ActiveRecord::Schema.define do                                                   
  create_table :posts, force: true do |t|                                        
  end                                                                            
                                                                                 
  create_table :comments, force: true do |t|                                     
    t.integer :post_id                                                           
  end                                                                            
end                                                                              
                                                                                 
class Post < ActiveRecord::Base                                                  
  has_many :comments                                                             
end                                                                              
                                                                                 
class Comment < ActiveRecord::Base                                               
  belongs_to :post                                                               
end                                                                              
                                                                                 
class BugTest < Minitest::Test                                                   
  def test_select_order_with_join                                                
    10.times { |i| Post.create! id: i + 1 }                                      
                                                                                 
    Post.first.comments << Comment.create!                                       
                                                                                 
    join_last = Post.left_joins(:comments).select "posts.*, comments.id"         
    join_first = Post.left_joins(:comments).select "comments.id, posts.*"        
                                                                                 
    assert_equal (1..10).to_a, join_last.map(&:id)                               
    assert_equal join_last.map(&:id), join_first.map(&:id)                       
  end                                                                            
end                                                                              
```

### Expected behavior
- Post records should have correct `id` and not the ID of the join record(s)
- Order of what is being selected should not matter?

### Actual behavior
The ID if the join record appears to be assigned to the main record

### System configuration
**Rails version**: 7.0.0

**Ruby version**: 3.2.2

Also tested with Rails 6 and Ruby 2.7

**Comment by Flixt** [comment]
TL; DR; 

I was able to verify this behavior. At least for me it was never the way I though `select` would work, but the confusion/issue is totally understandable when one come from a more SQLy point of view. Not sure this is a bug though.

---

I understand/share the confusion here, but I also think this is not how the `select` method is supposed to work (at least currently it is not). The arguments to `select` are, so to speak, mapped into the retrieved rows/records. 

E.g.

```ruby
10.times { |i| Post.create! id: i + 10 }

Post.first.comments << Comment.create!
Post.first.comments << Comment.create!

join_with_select = Post.left_joins(:comments).select("comments.id as comment_id, posts.*")
pp join_with_select.to_a.map { |p| [p.id, p.comment_id] }
```

would give something like (I let start the Post ids from 10 so one can spot the difference):

```ruby
[[10, 1],
 [10, 2],
 [11, nil],
 [12, nil],
 [13, nil],
 [14, nil],
 [15, nil],
 [16, nil],
 [17, nil],
 [18, nil],
 [19, nil]]
```

When using `Post.left_joins(:comments).select("posts.*, comments.id"), ActiveRecord will retrieve the columns
`id, id` and the second `id` (from the comments table) is used as `id` attribute for the Post model (`init_with_attributes` is used under the hood).

**Comment by fatkodima** [comment]
@Flixt is correct about the workings. 

I do not think this issue is fixable. Better to use a column alias, as suggested. Probably it is a good idea to update the docs to mention a similar scenario.
